<a href="https://colab.research.google.com/github/werowe/HypatiaAcademy/blob/master/ml/MultiOutputClassifier_LogisticRegression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np

df = pd.read_csv("https://raw.githubusercontent.com/Nazar-Popovych/Personal/refs/heads/main/StudentsPerformance.csv")
display(df)

gender race/ethnicity parental level of education         lunch  \
0    female        group B           bachelor's degree      standard   
1    female        group C                some college      standard   
2    female        group B             master's degree      standard   
3      male        group A          associate's degree  free/reduced   
4      male        group C                some college      standard   
..      ...            ...                         ...           ...   
995  female        group E             master's degree      standard   
996    male        group C                 high school  free/reduced   
997  female        group C                 high school  free/reduced   
998  female        group D                some college      standard   
999  female        group D                some college  free/reduced   

    test preparation course  math score  reading score  writing score  
0                      none          72             72             74  
1                 completed          69             90             88  
2                      none          90             95             93  
3                      none          47             57             44  
4                      none          76             78             75  
..                      ...         ...            ...            ...  
995               completed          88             99             95  
996                    none          62             55             55  
997               completed          59             71             65  
998               completed          68             78             77  
999                    none          77             86             86  

[1000 rows x 8 columns]

In [ ]:
df_stat=df[['gender', 'race/ethnicity', 'parental level of education', 'test preparation course', 'math score', 'reading score', 'writing score']]
df_encoded = pd.get_dummies(df_stat, drop_first=True, dtype=int)
display(df_encoded)

math score  reading score  writing score  gender_male  \
0            72             72             74            0   
1            69             90             88            0   
2            90             95             93            0   
3            47             57             44            1   
4            76             78             75            1   
..          ...            ...            ...          ...   
995          88             99             95            0   
996          62             55             55            1   
997          59             71             65            0   
998          68             78             77            0   
999          77             86             86            0   

     race/ethnicity_group B  race/ethnicity_group C  race/ethnicity_group D  \
0                         1                       0                       0   
1                         0                       1                       0   
2                         1                       0                       0   
3                         0                       0                       0   
4                         0                       1                       0   
..                      ...                     ...                     ...   
995                       0                       0                       0   
996                       0                       1                       0   
997                       0                       1                       0   
998                       0                       0                       1   
999                       0                       0                       1   

     race/ethnicity_group E  parental level of education_bachelor's degree  \
0                         0                                              1   
1                         0                                              0   
2                         0                                              0   
3                         0                                              0   
4                         0                                              0   
..                      ...                                            ...   
995                       1                                              0   
996                       0                                              0   
997                       0                                              0   
998                       0                                              0   
999                       0                                              0   

     parental level of education_high school  \
0                                          0   
1                                          0   
2                                          0   
3                                          0   
4                                          0   
..                                       ...   
995                                        0   
996                                        1   
997                                        1   
998                                        0   
999                                        0   

     parental level of education_master's degree  \
0                                              0   
1                                              0   
2                                              1   
3                                              0   
4                                              0   
..                                           ...   
995                                            1   
996                                            0   
997                                            0   
998                                            0   
999                                            0   

     parental level of education_some college  \
0                                           0   
1                                           1   
2                        

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.multioutput import MultiOutputClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

df_encoded['math_passed'] = (df_encoded['math score'] >= 60).astype(int)
df_encoded['reading_passed'] = (df_encoded['reading score'] >= 60).astype(int)
df_encoded['writing_passed'] = (df_encoded['writing score'] >= 60).astype(int)

X = df_encoded[['gender_male',
                'race/ethnicity_group B',	'race/ethnicity_group C',	'race/ethnicity_group D',	'race/ethnicity_group E',
                "parental level of education_bachelor's degree",	'parental level of education_high school',	"parental level of education_master's degree",	'parental level of education_some college',	'parental level of education_some high school',
                'test preparation course_none']]
y = df_encoded[['math_passed', 'reading_passed', 'writing_passed']]



X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

reg = MultiOutputClassifier(LogisticRegression(max_iter=1000))
reg.fit(X_train, y_train)
predictions = reg.predict(X_test)

print(classification_report(y_test, predictions))

              precision    recall  f1-score   support

           0       0.67      0.92      0.77       130
           1       0.72      0.98      0.83       142
           2       0.73      0.94      0.82       137

   micro avg       0.71      0.95      0.81       409
   macro avg       0.71      0.95      0.81       409
weighted avg       0.71      0.95      0.81       409
 samples avg       0.70      0.74      0.70       409



/usr/local/lib/python3.11/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Recall is ill-defined and being set to 0.0 in samples with no true labels. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


              precision    recall  f1-score   support

           0       0.67      0.92      0.77       130
           1       0.72      0.98      0.83       142
           2       0.73      0.93      0.82       137

   micro avg       0.71      0.94      0.81       409
   macro avg       0.71      0.94      0.81       409
weighted avg       0.71      0.94      0.81       409
 samples avg       0.70      0.74      0.70       409

#Per-Class Metrics

#Class 0
##Precision: 0.67
When the model predicts class 0, 67% of those predictions are correct (i.e., 33% are false positives).

##Recall: 0.92
The model correctly identifies 92% of all true class 0 samples (i.e., only 8% are missed).

##F1-score: 0.77
The harmonic mean of precision and recall for class 0.

#Class 1
##Precision: 0.72
When the model predicts class 1, 72% of those predictions are correct.

##Recall: 0.98
The model correctly identifies 98% of all true class 1 samples (very high recall).

##F1-score: 0.83
The harmonic mean of precision and recall for class 1.

#Class 2
##Precision: 0.73
When the model predicts class 2, 73% of those predictions are correct.

##Recall: 0.93
The model correctly identifies 93% of all true class 2 samples.

##F1-score: 0.82
The harmonic mean of precision and recall for class 2.

##Interpretation:
Your model is particularly strong in recall for all classes (especially class 1), meaning it finds most of the true instances. Precision is a bit lower, especially for class 0, meaning there are some false positives.



The meaning of the row number:                                                  
'0' is 'math_passed'                                                            
'1' is 'reading_passed'              
'2' is 'writing_passed'

In [ ]:
from statsmodels.stats.outliers_influence import variance_inflation_factor
from statsmodels.tools.tools import add_constant

X = add_constant(X)

vif = pd.DataFrame()
vif["Feature"] = X.columns
vif["VIF"] = [variance_inflation_factor(X.values, i) for i in range(X.shape[1])]
print(vif)

                                          Feature        VIF
0                                           const  18.776538
1                                     gender_male   1.012923
2                          race/ethnicity_group B   2.562508
3                          race/ethnicity_group C   3.171347
4                          race/ethnicity_group D   2.938888
5                          race/ethnicity_group E   2.245809
6   parental level of education_bachelor's degree   1.353583
7         parental level of education_high school   1.528770
8     parental level of education_master's degree   1.201905
9        parental level of education_some college   1.568481
10   parental level of education_some high school   1.505750
11                   test preparation course_none   1.015661
